backend.go

```diff
@@ -800,50 +808,110 @@ type DecodeData struct {
+type RawFragment = []byte
+type ValidatedFragment = []byte
+
+func (backend *Backend) ValidateFragmentMatrix(frag RawFragment, pieceSize int) bool {
+	result := C.check_matrix_fragment((*C.char)(unsafe.Pointer(&frag[0])), C.int(len(frag)), C.int(pieceSize))
+	if result != C.bool(true) {
+		return false
+	}
+
+	return true
+}
+
+func (backend *Backend) LinearizeMatrix(frags []ValidatedFragment, pieceSize int) (*DecodeData, error) {
 	var wg sync.WaitGroup
 
 	if len(frags) == 0 {
-		return nil, errors.New("decoding requires at least one fragment")
+		return nil, errors.New("linearizing requires at least one fragment")
 	}
 
-	fragLen := len(frags[0])
-	lenBlock := piecesize + backend.headerSize
-	numBlock := fragLen / lenBlock
-	if numBlock*lenBlock != fragLen {
-		numBlock++
+	fragRangeLen := len(frags[0])
+	chunkSize := pieceSize + backend.headerSize
+	nrChunks := fragRangeLen / chunkSize
+	if nrChunks*chunkSize != fragRangeLen {
+		nrChunks++
 	}
 
-	// allocate output buffer
-	dataB, data := backend.pool.New(numBlock * piecesize * backend.K)
+	/* Fragments are sorted beforehand with the index of the first chunk.
+	   All chunks of a fragments share the same index. */
+	fragsIndex := make([]int, len(frags))
+	for i := 0; i < len(frags); i += 1 {
+		fragsIndex[i] = i
+	}
+
+	sort.Slice(fragsIndex, func(i, j int) bool {
+		lhs := frags[fragsIndex[i]]
+		rhs := frags[fragsIndex[j]]
+
+		var lhsIdx, rhsIdx C.int
+		lhsIdx = C.get_fragment_idx((*C.char)(unsafe.Pointer(&lhs[0])))
+		rhsIdx = C.get_fragment_idx((*C.char)(unsafe.Pointer(&rhs[0])))
+
+		return lhsIdx < rhsIdx
+	})
+
+	/* There is no reconstruction that can happen.
+	   All coding fragments must be ignored */
+	lastDataFragIdxExcl := 0
+	previousFragIdx := -1
+	for lastDataFragIdxExcl < len(fragsIndex) {
+		frag := frags[fragsIndex[lastDataFragIdxExcl]]
+		idx := int(C.get_fragment_idx((*C.char)(unsafe.Pointer(&frag[0]))))
+
+		if idx < 0 {
+			return nil, errors.New("invalid fragment index")
+		}
 
+		if idx >= backend.K {
+			break
+		}
+
+		if len(frags[0]) != fragRangeLen {
+			return nil, errors.New("invalid fragment len")
+		}
+
+		if previousFragIdx > 0 && (idx-previousFragIdx) != 1 {
+			/* Fragments are not contiguous. This functions doesn't supports
+			   gaps. */
+			return nil, errors.New("gaps in the provided fragments")
+		}
+
+		lastDataFragIdxExcl += 1
+		previousFragIdx = idx
+	}
+	fragsIndex = fragsIndex[:lastDataFragIdxExcl]
+
+	dataB, data := backend.pool.New(nrChunks * pieceSize * len(fragsIndex))
 	errorNb := uint32(0)
 	totLen := uint64(0)
-	wg.Add(numBlock)
+	wg.Add(nrChunks)
 
-	for i := 0; i < numBlock; i++ {
+	for i := 0; i < nrChunks; i++ {
 		// launch goroutines, providing them a subrange of the final buffer so it can be used
 		// in concurrency without need to lock it access
-		go func(blockNr int) {
-			cFrags := C.makeStrArray(C.int(len(frags)))
+		func(chunkIdx int) {
```

Comment:
We don't launch goroutines anymore?

Reply:
A rebase mishap (for debugging). Should be fixed.